BZ2093243 Improve baremetal master node replacement documentation (#47751)
openshift/openshift-docs: modules/restore-replace-stopped-baremetal-etcd-member.adoc

// Module included in the following assemblies:
//
// * /backup_and_restore/control_plane_backup_and_restore/replacing-unhealthy-etcd-member.adoc

:_content-type: PROCEDURE
[id="restore-replace-stopped-baremetal-etcd-member_{context}"]
= Replacing an unhealthy bare metal etcd member whose machine is not running or whose node is not ready

This procedure details the steps to replace a bare metal etcd member that is unhealthy either because the machine is not running or because the node is not ready.

.Prerequisites

* You have identified the unhealthy bare metal etcd member.
* You have verified that either the machine is not running or the node is not ready.
* You have access to the cluster as a user with the `cluster-admin` role.
* You have taken an etcd backup.
+
[IMPORTANT]
====
It is important to take an etcd backup before performing this procedure so that your cluster can be restored if you encounter any issues.
====

.Procedure

. Remove the unhealthy member.

.. Choose a pod that is _not_ on the affected node:
+
In a terminal that has access to the cluster as a `cluster-admin` user, run the following command:
+
[source,terminal]
----

> $ oc get pods -n openshift-etcd | grep -v etcd-quorum-guard | grep etcd

Reviewer: $ oc get pods -n openshift-etcd | grep -v etcd-guard| grep etcd

Author: Done

----
+
.Example output
[source,terminal]
----
etcd-ip-10-0-131-183.ec2.internal                3/3     Running     0          123m
etcd-ip-10-0-164-97.ec2.internal                 3/3     Running     0          123m

> etcd-ip-10-0-154-204.ec2.internal                3/3     Running     0          124m

Reviewer: etcd-openshift-master-0                5/5     Running     0          105m
etcd-openshift-master-1                5/5     Running     0          107m
etcd-openshift-master-2                5/5     Running     0          103m

Author: Done

----

.. Connect to the running etcd container, passing in the name of a pod that is not on the affected node:
+
In a terminal that has access to the cluster as a `cluster-admin` user, run the following command:
+
[source,terminal]
----

> $ oc rsh -n openshift-etcd etcd-ip-10-0-154-204.ec2.internal

Reviewer: $ oc rsh -n openshift-etcd etcd-openshift-master-0

Author: Done

----

.. View the member list:
+
[source,terminal]
----
sh-4.2# etcdctl member list -w table
----
+
.Example output
[source,terminal]
----
+------------------+---------+------------------------------+---------------------------+---------------------------+
|        ID        | STATUS  |             NAME             |        PEER ADDRS         |       CLIENT ADDRS        |
+------------------+---------+------------------------------+---------------------------+---------------------------+
| 6fc1e7c9db35841d | started | ip-10-0-131-183.ec2.internal | https://10.0.131.183:2380 | https://10.0.131.183:2379 |
| 757b6793e2408b6c | started |  ip-10-0-164-97.ec2.internal |  https://10.0.164.97:2380 |  https://10.0.164.97:2379 |
| ca8c2990a0aa29d1 | started | ip-10-0-154-204.ec2.internal | https://10.0.154.204:2380 | https://10.0.154.204:2379 |
+------------------+---------+------------------------------+---------------------------+---------------------------+

> ----

Reviewer: +------------------+---------+--------------------+---------------------------+---------------------------+------------+
|        ID        | STATUS  |        NAME        |        PEER ADDRS         |       CLIENT ADDRS        | IS LEARNER |
+------------------+---------+--------------------+---------------------------+---------------------------+------------+
| 7a8197040a5126c8 | started | openshift-master-2 | https://10.46.29.131:2380/ | https://10.46.29.131:2379/ |      false |
| 8d5abe9669a39192 | started | openshift-master-1 | https://10.46.29.130:2380/ | https://10.46.29.130:2379/ |      false |
| cc3830a72fc357f9 | started | openshift-master-0 | https://10.46.29.129:2380/ | https://10.46.29.129:2379/ |      false |
+------------------+---------+--------------------+---------------------------+---------------------------+------------+

Author: Done

+
Take note of the ID and the name of the unhealthy etcd member, because these values are needed later in the procedure. The `$ etcdctl endpoint health` command will list the removed member until the procedure of replacement is finished and a new member is added.

> .. Remove the unhealthy etcd member by providing the ID to the `etcdctl member remove` command:

Reviewer: This part should be moved to after bmh deletion

Author: @lshilin-r It appears the problem is with the sentence "The `$ etcdctl endpoint health` command will list the removed member until the procedure of replacement is finished and a new member is added.", is this correct? If so, move only this sentence to after bmh deletion?

Author: @lshilin-r Please forgive me, I've not moved this yet - I need more clarification.

Reviewer: No need to move it. I thought I mentioned it here, but may be did it on slack only

+
[source,terminal]
----

> sh-4.2# etcdctl member remove 6fc1e7c9db35841d

Reviewer: sh-4.4# etcdctl member remove 7a8197040a5126c8

Author: Done

----
+
.Example output
[source,terminal]
----

> Member 6fc1e7c9db35841d removed from cluster ead669ce1fbfb346

Reviewer: Member 7a8197040a5126c8 removed from cluster b23536c33f2cdd1b

Author: Done

+
[source,terminal]
----
$ oc get machine clustername-8qw5l-master-0 \ <1>
    -n openshift-machine-api \
    -o yaml \
    > new-master-machine.yaml
----
<1> Specify the name of the control plane machine for the unhealthy node.

.. Edit the `new-master-machine.yaml` file that was created in the previous step to assign a new name and remove unnecessary fields.

... Remove the entire `status` section:
+
[source,yaml]
----
status:
  addresses:
  - address: 10.0.131.183
    type: InternalIP
  - address: ip-10-0-131-183.ec2.internal
    type: InternalDNS
  - address: ip-10-0-131-183.ec2.internal
    type: Hostname
  lastUpdated: "2020-04-20T17:44:29Z"
  nodeRef:
    kind: Node
    name: ip-10-0-131-183.ec2.internal
    uid: acca4411-af0d-4387-b73e-52b2484295ad
  phase: Running
  providerStatus:
    apiVersion: awsproviderconfig.openshift.io/v1beta1
    conditions:
    - lastProbeTime: "2020-04-20T16:53:50Z"
      lastTransitionTime: "2020-04-20T16:53:50Z"
      message: machine successfully created
      reason: MachineCreationSucceeded
      status: "True"
      type: MachineCreation
    instanceId: i-0fdb85790d76d0c3f
    instanceState: stopped

>     kind: AWSMachineProviderStatus

Reviewer: status:
  addresses:
  - address: ""
    type: InternalIP
  - address: fe80::4adf:37ff:feb0:8aa1%ens1f1.373
    type: InternalIP
  - address: fe80::4adf:37ff:feb0:8aa1%ens1f1.371
    type: InternalIP
  - address: fe80::4adf:37ff:feb0:8aa1%ens1f1.341
    type: InternalIP
  - address: ""
    type: InternalIP
  - address: 10.46.29.131
    type: InternalIP
  - address: fe80::4adf:37ff:feb0:8aa1%ens1f1
    type: InternalIP
  - address: fe80::4adf:37ff:feb0:8aa0%ens1f0.342
    type: InternalIP
  - address: fe80::4adf:37ff:feb0:8aa1%ens1f1.372
    type: InternalIP
  - address: fe80::4adf:37ff:feb0:8aa1%ens1f1.370
    type: InternalIP
  - address: ""
    type: InternalIP
  - address: 16.1.15.2
    type: InternalIP
  - address: fe80::64f7:72ff:feec:f155%enp1s0f4u4
    type: InternalIP
  - address: openshift-master-2
    type: Hostname
  - address: openshift-master-2
    type: InternalDNS

Author: done

Author: @lshilin-r I used only a subset of this output - do you prefer I use ALL of it?

----

... Change the `metadata.name` field to a new name.
+
It is recommended to keep the same base name as the old machine and change the ending number to the next available number. In this example, `clustername-8qw5l-master-0` is changed to `clustername-8qw5l-master-3`.
+
For example:
+
[source,yaml]
----
apiVersion: machine.openshift.io/v1beta1
kind: Machine
metadata:
  ...

>   name: clustername-8qw5l-master-3

Reviewer: name: ocp-edge-4zg89-master-3

Author: done

  ...
----

... Remove the `spec.providerID` field:
+
[source,yaml]
----

>   providerID: aws:///us-east-1a/i-0fdb85790d76d0c3f

Reviewer: providerID: baremetalhost:///openshift-machine-api/openshift-master-2/3354bdac-61d8-410f-be5b-6a395b056135

Author: Done

----

... Remove the `metadata.annotations` and `metadata.generation` fields:
+
[source,yaml]
----
  annotations:

>     machine.openshift.io/instance-state: running

Reviewer: machine.openshift.io/instance-state: externally provisioned

Author: done

  ...
  generation: 2
----

... Remove the `metadata.resourceVersion` and `metadata.uid` fields:
+
[source,yaml]
----
  resourceVersion: "13291"
  uid: a282eb70-40a2-4e89-8009-d05dd420d31a
----

> 

Reviewer: Remove the `spec.conditions`,  `spec.lastUpdated`,  `spec.nodeRef` and `spec.phase`

  conditions:
  - lastTransitionTime: "2022-08-03T08:40:36Z"
    message: 'Drain operation currently blocked by: [{Name:EtcdQuorumOperator Owner:clusteroperator/etcd}]'
    reason: HookPresent
    severity: Warning
    status: "False"
    type: Drainable
  - lastTransitionTime: "2022-08-03T08:39:55Z"
    status: "True"
    type: InstanceExists
  - lastTransitionTime: "2022-08-03T08:36:37Z"
    status: "True"
    type: Terminable
  lastUpdated: "2022-08-03T08:40:36Z"
  nodeRef:
    kind: Node
    name: openshift-master-2
    uid: 788df282-6507-4ea2-9a43-24f237ccbc3c
  phase: Running

Author: done

.. Delete the machine of the unhealthy member:
+
[source,terminal]
----
$ oc delete machine -n openshift-machine-api clustername-8qw5l-master-0 <1>
----
<1> Specify the name of the control plane machine for the unhealthy node.

> 

Author: Thank you so much @lshilin-r for this important information. I will work to revise the document and incorporate what you have provided.

> 

Author: Done

.. Verify that the machine was deleted:
+
[source,terminal]
----
$ oc get machines -n openshift-machine-api -o wide
----
+
.Example output
[source,terminal]
----
NAME                                        PHASE     TYPE        REGION      ZONE         AGE     NODE                           PROVIDERID                              STATE
clustername-8qw5l-master-1                  Running   m4.xlarge   us-east-1   us-east-1b   3h37m   ip-10-0-154-204.ec2.internal   aws:///us-east-1b/i-096c349b700a19631   running
clustername-8qw5l-master-2                  Running   m4.xlarge   us-east-1   us-east-1c   3h37m   ip-10-0-164-97.ec2.internal    aws:///us-east-1c/i-02626f1dba9ed5bba   running
clustername-8qw5l-worker-us-east-1a-wbtgd   Running   m4.large    us-east-1   us-east-1a   3h28m   ip-10-0-129-226.ec2.internal   aws:///us-east-1a/i-010ef6279b4662ced   running
clustername-8qw5l-worker-us-east-1b-lrdxb   Running   m4.large    us-east-1   us-east-1b   3h28m   ip-10-0-144-248.ec2.internal   aws:///us-east-1b/i-0cb45ac45a166173b   running

> clustername-8qw5l-worker-us-east-1c-pkg26   Running   m4.large    us-east-1   us-east-1c   3h28m   ip-10-0-170-181.ec2.internal   aws:///us-east-1c/i-06861c00007751b0a   running

Reviewer: $ oc get machines -n openshift-machine-api -o wide
NAME                            PHASE     TYPE   REGION   ZONE   AGE     NODE                 PROVIDERID                                                                                       STATE
ocp-edge-4zg89-master-0         Running                          3h39m   openshift-master-0   baremetalhost:///openshift-machine-api/openshift-master-0/da1ebe11-3ff2-41c5-b099-0aa41222964e   externally provisioned
ocp-edge-4zg89-master-1         Running                          3h39m   openshift-master-1   baremetalhost:///openshift-machine-api/openshift-master-1/d9f9acbc-329c-475e-8d81-03b20280a3e1   externally provisioned
ocp-edge-4zg89-worker-0-8v2jj   Running                          3h13m   openshift-worker-0   baremetalhost:///openshift-machine-api/openshift-worker-0/3d685b81-7410-4bb3-80ec-13a31858241f   provisioned
ocp-edge-4zg89-worker-0-jb5z9   Running                          3h13m   openshift-worker-1   baremetalhost:///openshift-machine-api/openshift-worker-1/0fdae6eb-2066-4241-91dc-e7ea72ab13b9   provisioned

Author: Done

----

[IMPORTANT]
====
If you reuse the `BareMetalHost` object definition from an existing control plane host, do not leave the `externallyProvisioned` field set to `true`.

Existing control plane `BareMetalHost` objects may have the `externallyProvisioned` flag set to `true` if they were provisioned by the {product-title} installation program.
====

> .Prerequisites

Reviewer: shouldn't this part be the first

Author: Yes! I have moved. Thank you!

* You have access to the cluster as a user with the `cluster-admin` role.

* You have taken an etcd backup.
+
[IMPORTANT]
====
Take an etcd backup before performing this procedure so that you can restore your cluster if you encounter any issues. For more information about taking an etcd backup, see the _Additional resources_ section.
====

.Procedure

. Ensure that the Bare Metal Operator is available:
+
[source,terminal]
----
$ oc get clusteroperator baremetal
----
+
.Example output
[source,terminal]
----
NAME        VERSION   AVAILABLE   PROGRESSING   DEGRADED   SINCE   MESSAGE

> baremetal   4.10.12   True        False         False      3d15h

Author: This document fixes a problem found in 4.10.12 and identified by telco customer NOKIA. If I edit a number just to make it look like it was run on 4.11, that might raise questions about fidelity.

> baremetal   4.10.12   True        False         False      3d15h

Reviewer: ok. May be u can use something like <version>?

> baremetal   4.10.12   True        False         False      3d15h

Author: @lshilin-r May I use the version of OpenShift that you yourself tested on?

> baremetal   4.10.12   True        False         False      3d15h

Author: Done

----

> . Remove the old `BareMetalHost` and `Machine` objects:

Reviewer: ```suggestion
. Remove the old `BareMetalHost` object:
```

Author: Done

+
[source,terminal]
----
$ oc delete bmh -n openshift-machine-api vmaster-0

> $ oc delete machine -n openshift-machine-api kni1-master-0

Author: We did! Great catch, @lshilin-r ! Thank you so much!

> $ oc delete machine -n openshift-machine-api kni1-master-0

Reviewer: BTW, looks like this examples are coming from different setup (see the name of the machine). In addition, the name of bmh is very confusing, it looks like it comes from virtual simulation.

I would delete BMH first. To find the list of BMH the user need to run a command:
$ oc get bmh -n openshift-machine-api
The BMH to delete has a consumer name of the machine to delete

> $ oc delete machine -n openshift-machine-api kni1-master-0

Author: Agreed, thank you. I am glad to add the "oc get bmh ... "

> $ oc delete machine -n openshift-machine-api kni1-master-0

Reviewer: $ oc get bmh -n openshift-machine-api
NAME                 STATE                    CONSUMER                        ONLINE   ERROR   AGE
openshift-master-0   externally provisioned   ocp-edge-4zg89-master-0         true             3h43m
openshift-master-1   externally provisioned   ocp-edge-4zg89-master-1         true             3h43m
openshift-master-2   externally provisioned   ocp-edge-4zg89-master-2         true             3h43m
openshift-worker-0   provisioned              ocp-edge-4zg89-worker-0-8v2jj   true             3h43m
openshift-worker-1   provisioned              ocp-edge-4zg89-worker-0-jb5z9   true             3h43m

> $ oc delete machine -n openshift-machine-api kni1-master-0

Reviewer: ```suggestion
$ oc delete bmh openshift-master-2 -n openshift-machine-api
baremetalhost.metal3.io "openshift-master-2" deleted
```

Reviewer: please, don't forget to remove `delete machine`

Author: Done

> $ oc delete machine -n openshift-machine-api kni1-master-0

Author: done

----
+

> After you remove the `BareMetalHost` and `Machine` objects, then the `Machine` controller automatically deletes the `Node` object.

Reviewer: sometimes it takes a little long for machine controller to realize the bmh and machine deleted. In this case the node can be deleted by
$ oc delete node openshift-master-2
node "openshift-master-2" deleted

$ oc get nodes
NAME                 STATUS   ROLES    AGE     VERSION
openshift-master-0   Ready    master   3h24m   v1.24.0+9546431
openshift-master-1   Ready    master   3h24m   v1.24.0+9546431
openshift-worker-0   Ready    worker   176m    v1.24.0+9546431
openshift-worker-1   Ready    worker   176m    v1.24.0+9546431

Author: Done. Thank you for this wonderful addition to the document!

. Create the new `BareMetalHost` object and the secret to store the BMC credentials:
+
[source,terminal]
----
$ cat <<EOF | oc apply -f -
apiVersion: v1
kind: Secret
metadata:
  name: kni1-master-0-bmc-secret
  namespace: openshift-machine-api
data:
  password: <username>
  username: <password>
type: Opaque
---
apiVersion: metal3.io/v1alpha1
kind: BareMetalHost
metadata:
  name: kni1-master-0
  namespace: openshift-machine-api
spec:
  automatedCleaningMode: disabled
  bmc:

>     address: redfish-virtualmedia+http://192.168.124.113:8000/redfish/v1/Systems/f87eaf82-b32d-4291-ace6-b28677964e78

Author: Thank you so much. Please tell me again, is the spelling of the third protocol "idrac-vistualmedia" or "idrac-visualmedia"?

>     address: redfish-virtualmedia+http://192.168.124.113:8000/redfish/v1/Systems/f87eaf82-b32d-4291-ace6-b28677964e78

Reviewer: idrac-vistualmedia

>     address: redfish-virtualmedia+http://192.168.124.113:8000/redfish/v1/Systems/f87eaf82-b32d-4291-ace6-b28677964e78

Author: and the "oc get bmh ... " will give the user that information, right?

>     address: redfish-virtualmedia+http://192.168.124.113:8000/redfish/v1/Systems/f87eaf82-b32d-4291-ace6-b28677964e78

Author: Done!

    credentialsName: kni1-master-0-bmc-secret
  bootMACAddress: aa:aa:aa:aa:ab:03
  bootMode: UEFI
  externallyProvisioned: false
  hardwareProfile: unknown

>   online: true

Reviewer: apiVersion: v1
kind: Secret
metadata:
  name: openshift-master-2-bmc-secret
  namespace: openshift-machine-api
data:
  password: <password>
  username: <username>
type: Opaque
---
apiVersion: metal3.io/v1alpha1
kind: BareMetalHost
metadata:
  name: openshift-master-2
  namespace: openshift-machine-api
spec:
  automatedCleaningMode: disabled
  bmc:
    address: redfish-virtualmedia://10.46.61.18:443/redfish/v1/Systems/1
    credentialsName: openshift-master-2-bmc-secret
    disableCertificateVerification: true
  bootMACAddress: 48:df:37:b0:8a:a0
  bootMode: UEFI
  externallyProvisioned: false
  online: true
  rootDeviceHints:
    deviceName: /dev/sda
  userData:
    name: master-user-data-managed
    namespace: openshift-machine-api

Author: Done

Author: Also: added the Note:

EOF
----
+
After the inspection is complete, the `BareMetalHost` object is created and available to be provisioned.

. View available `BareMetalHost` objects:
+[source,terminal]
----
$ oc get bmh -n openshift-machine-api
----
+
.Example output
[source,terminal]
----
NAME           STATE                    CONSUMER                   ONLINE   ERROR   AGE
kni1-master-0  available                ocp-hkw9p-master-0         true             1h10m
kni1-master-1  externally provisioned   ocp-hkw9p-master-1         true             4h53m
kni1-master-2  externally provisioned   ocp-hkw9p-master-2         true             4h53m
kni1-worker-0  provisioned              ocp-hkw9p-worker-0-ktmmx   true             4h53m

> kni1-worker-1  provisioned              ocp-hkw9p-worker-0-l2zmb   true             4h53m

Reviewer: on this stage the new BMH have no consumer yet. 
$ oc get bmh -n openshift-machine-api
NAME                 STATE                    CONSUMER                        ONLINE   ERROR   AGE
openshift-master-0   externally provisioned   ocp-edge-4zg89-master-0         true             4h6m
openshift-master-1   externally provisioned   ocp-edge-4zg89-master-1         true             4h6m
openshift-master-2   available                                               true             35m14s
openshift-worker-0   provisioned              ocp-edge-4zg89-worker-0-8v2jj   true             4h6m
openshift-worker-1   provisioned              ocp-edge-4zg89-worker-0-jb5z9   true             4h6m

Author: Done

clustername-8qw5l-master-3                  Provisioning   m4.xlarge   us-east-1   us-east-1a   85s     ip-10-0-133-53.ec2.internal    aws:///us-east-1a/i-015b0888fe17bc2c8   running <1>
clustername-8qw5l-worker-us-east-1a-wbtgd   Running        m4.large    us-east-1   us-east-1a   3h28m   ip-10-0-129-226.ec2.internal   aws:///us-east-1a/i-010ef6279b4662ced   running
clustername-8qw5l-worker-us-east-1b-lrdxb   Running        m4.large    us-east-1   us-east-1b   3h28m   ip-10-0-144-248.ec2.internal   aws:///us-east-1b/i-0cb45ac45a166173b   running
clustername-8qw5l-worker-us-east-1c-pkg26   Running        m4.large    us-east-1   us-east-1c   3h28m   ip-10-0-170-181.ec2.internal   aws:///us-east-1c/i-06861c00007751b0a   running
----
<1> The new machine, `clustername-8qw5l-master-3` is being created and is ready once the phase changes from `Provisioning` to `Running`.
+
It might take a few minutes for the new machine to be created. The etcd cluster Operator will automatically sync when the machine or node returns to a healthy state.

> 

Reviewer: $ oc get bmh -n openshift-machine-api
NAME                 STATE                    CONSUMER                        ONLINE   ERROR   AGE
openshift-master-0   externally provisioned   ocp-edge-4zg89-master-0         true             4h48m
openshift-master-1   externally provisioned   ocp-edge-4zg89-master-1         true             4h48m
openshift-master-2   provisioned              ocp-edge-4zg89-master-3         true             47m
openshift-worker-0   provisioned              ocp-edge-4zg89-worker-0-8v2jj   true             4h48m
openshift-worker-1   provisioned              ocp-edge-4zg89-worker-0-jb5z9   true             4h48m
[kni@ocp-edge06 ~]$ oc get machine -n openshift-machine-api
NAME                            PHASE     TYPE   REGION   ZONE   AGE
ocp-edge-4zg89-master-0         Running                          4h49m
ocp-edge-4zg89-master-1         Running                          4h48m
ocp-edge-4zg89-master-3         Running                          25m
ocp-edge-4zg89-worker-0-8v2jj   Running                          4h23m
ocp-edge-4zg89-worker-0-jb5z9   Running                          4h23m
[kni@ocp-edge06 ~]$ oc get nodes
NAME                 STATUS   ROLES    AGE     VERSION
openshift-master-0   Ready    master   4h26m   v1.24.0+9546431
openshift-master-1   Ready    master   4h26m   v1.24.0+9546431
openshift-master-2   Ready    master   12m     v1.24.0+9546431
openshift-worker-0   Ready    worker   3h58m   v1.24.0+9546431
openshift-worker-1   Ready    worker   3h58m   v1.24.0+9546431

Reviewer: did you decide not to give this examples of the expected cluster state?

Author: I'm glad to do so but formatting of the output text is challenging.....

Author: Done

Reviewer: It looks like it added not to the right place (see my previous comment)

.Verification

. Verify that all etcd pods are running properly.
+
In a terminal that has access to the cluster as a `cluster-admin` user, run the following command:
+
[source,terminal]
----

> $ oc get pods -n openshift-etcd | grep -v etcd-quorum-guard | grep etcd

Reviewer: $ oc get pods -n openshift-etcd | grep -v etcd-guard| grep etcd

Author: done

----
+
.Example output
[source,terminal]
----
etcd-ip-10-0-133-53.ec2.internal                 3/3     Running     0          7m49s
etcd-ip-10-0-164-97.ec2.internal                 3/3     Running     0          123m
etcd-ip-10-0-154-204.ec2.internal                3/3     Running     0          124m

> ----

Reviewer: etcd-openshift-master-0                 5/5     Running     0          9m36s
etcd-openshift-master-1                 5/5     Running     0          11m
etcd-openshift-master-2                 5/5     Running     0          7m36s

Author: Done

+
If the output from the previous command only lists two pods, you can manually force an etcd redeployment. In a terminal that has access to the cluster as a `cluster-admin` user, run the following command:
+
[source,terminal]
----
$ oc patch etcd cluster -p='{"spec": {"forceRedeploymentReason": "recovery-'"$( date --rfc-3339=ns )"'"}}' --type=merge <1>
----
<1> The `forceRedeploymentReason` value must be unique, which is why a timestamp is appended.

. Verify that there are exactly three etcd members.

.. Connect to the running etcd container, passing in the name of a pod that was not on the affected node:
+
In a terminal that has access to the cluster as a `cluster-admin` user, run the following command:
+
[source,terminal]
----

> $ oc rsh -n openshift-etcd etcd-ip-10-0-154-204.ec2.internal

Reviewer: $ oc rsh -n openshift-etcd etcd-openshift-master-0

Author: Done

----

.. View the member list:
+
[source,terminal]
----
sh-4.2# etcdctl member list -w table
----
+
.Example output
[source,terminal]
----
+------------------+---------+------------------------------+---------------------------+---------------------------+
|        ID        | STATUS  |             NAME             |        PEER ADDRS         |       CLIENT ADDRS        |
+------------------+---------+------------------------------+---------------------------+---------------------------+
| 5eb0d6b8ca24730c | started |  ip-10-0-133-53.ec2.internal |  https://10.0.133.53:2380 |  https://10.0.133.53:2379 |
| 757b6793e2408b6c | started |  ip-10-0-164-97.ec2.internal |  https://10.0.164.97:2380 |  https://10.0.164.97:2379 |
| ca8c2990a0aa29d1 | started | ip-10-0-154-204.ec2.internal | https://10.0.154.204:2380 | https://10.0.154.204:2379 |

> +------------------+---------+------------------------------+---------------------------+---------------------------+

Reviewer: +------------------+---------+--------------------+---------------------------+---------------------------+------------+
|        ID        | STATUS  |        NAME        |        PEER ADDRS         |       CLIENT ADDRS        | IS LEARNER |
+------------------+---------+--------------------+---------------------------+---------------------------+------------+
| 8d5abe9669a39192 | started | openshift-master-1 | https://10.46.29.130:2380 | https://10.46.29.130:2379 |      false |
| cb1ce22c4574be06 | started | openshift-master-2 | https://10.46.29.131:2380 | https://10.46.29.131:2379 |      false |
| cc3830a72fc357f9 | started | openshift-master-0 | https://10.46.29.129:2380 | https://10.46.29.129:2379 |      false |
+------------------+---------+--------------------+---------------------------+---------------------------+------------+

Author: Done

----
+
If the output from the previous command lists more than three etcd members, you must carefully remove the unwanted member.
+
[WARNING]
====
Be sure to remove the correct etcd member; removing a good etcd member might lead to quorum loss.

> ====

Reviewer: the following verification can be added:
Validate that all etcd members are healthy 
```
sh-4.4# etcdctl endpoint health --cluster 
https://10.46.29.130:2379 is healthy: successfully committed proposal: took = 8.973065ms
https://10.46.29.129:2379 is healthy: successfully committed proposal: took = 11.559829ms
https://10.46.29.131:2379 is healthy: successfully committed proposal: took = 11.665203ms

```

Validate that all nodes are at the latest revision
```
$ oc get etcd -o=jsonpath='{range.items[0].status.conditions[?(@.type=="NodeInstallerProgressing")]}{.reason}{"\n"}{.message}{"\n"}'
AllNodesAtLatestRevision
```

Author: Done, thank you!

examplecluster-control-plane-0         Running                          3h11m   openshift-control-plane-0   baremetalhost:///openshift-machine-api/openshift-control-plane-0/da1ebe11-3ff2-41c5-b099-0aa41222964e   externally provisioned <1>
examplecluster-control-plane-1         Running                          3h11m   openshift-control-plane-1   baremetalhost:///openshift-machine-api/openshift-control-plane-1/d9f9acbc-329c-475e-8d81-03b20280a3e1   externally provisioned
examplecluster-control-plane-2         Running                          3h11m   openshift-control-plane-2   baremetalhost:///openshift-machine-api/openshift-control-plane-2/3354bdac-61d8-410f-be5b-6a395b056135   externally provisioned
examplecluster-compute-0   Running                          165m    openshift-compute-0   baremetalhost:///openshift-machine-api/openshift-compute-0/3d685b81-7410-4bb3-80ec-13a31858241f   provisioned
examplecluster-compute-1   Running                          165m    openshift-compute-1   baremetalhost:///openshift-machine-api/openshift-compute-1/0fdae6eb-2066-4241-91dc-e7ea72ab13b9   provisioned

> ----

Reviewer: 2 last lines problem with tabs

<1> This is the control plane machine for the unhealthy node, `examplecluster-control-plane-2`.

.. Save the machine configuration to a file on your file system:
+
[source,terminal]
----

> $ oc get machine clustername-8qw5l-master-2 \ <1>

Reviewer: ```suggestion
$ oc get machine examplecluster-control-plane-2 \ <1>
```

Author: Done! Thank you!

    -n openshift-machine-api \
    -o yaml \
    > new-master-machine.yaml
----
Specify the name of the control plane machine for the unhealthy node.

.. Edit the `new-master-machine.yaml` file that was created in the previous step to assign a new name and remove unnecessary fields.

... Remove the entire `status` section:
+
[source,yaml]
----
status:
  addresses:
  - address: ""
    type: InternalIP
  - address: fe80::4adf:37ff:feb0:8aa1%ens1f1.373
    type: InternalDNS
  - address: fe80::4adf:37ff:feb0:8aa1%ens1f1.371
    type: Hostname
  lastUpdated: "2020-04-20T17:44:29Z"
  nodeRef:
    kind: Machine
    name: fe80::4adf:37ff:feb0:8aa1%ens1f1.372
    uid: acca4411-af0d-4387-b73e-52b2484295ad
  phase: Running
  providerStatus:
    apiVersion: machine.openshift.io/v1beta1
    conditions:
    - lastProbeTime: "2020-04-20T16:53:50Z"
      lastTransitionTime: "2020-04-20T16:53:50Z"
      message: machine successfully created
      reason: MachineCreationSucceeded
      status: "True"
      type: MachineCreation
    instanceId: i-0fdb85790d76d0c3f
    instanceState: stopped
    kind: Machine
----

. Change the `metadata.name` field to a new name.
+

> It is recommended to keep the same base name as the old machine and change the ending number to the next available number. In this example, `clustername-8qw5l-master-0` is changed to `clustername-8qw5l-master-2`.

Reviewer: ```suggestion
It is recommended to keep the same base name as the old machine and change the ending number to the next available number. In this example, `examplecluster-control-plane-2` is changed to `examplecluster-control-plane-3`.
```

Author: Done, thank you!

+
[source,yaml]
----
  annotations:
    machine.openshift.io/instance-state: externally provisioned
  ...
  generation: 2
----

.. Remove the `spec.conditions`, `spec.lastUpdated`, `spec.nodeRef` and `spec.phase' fields:
+
[source,yaml]
----
  lastTransitionTime: "2022-08-03T08:40:36Z"
message: 'Drain operation currently blocked by: [{Name:EtcdQuorumOperator Owner:clusteroperator/etcd}]'
reason: HookPresent
severity: Warning
status: "False"

type: Drainable
lastTransitionTime: "2022-08-03T08:39:55Z"
status: "True"
type: InstanceExists

lastTransitionTime: "2022-08-03T08:36:37Z"
status: "True"
type: Terminable
lastUpdated: "2022-08-03T08:40:36Z"
nodeRef:
kind: Node
name: openshift-control-plane-2
uid: 788df282-6507-4ea2-9a43-24f237ccbc3c
phase: Running
----

. Ensure that the Bare Metal Operator is available:
+
[source,terminal]
----
$ oc get clusteroperator baremetal
----
+
.Example output
[source,terminal]
----
NAME        VERSION   AVAILABLE   PROGRESSING   DEGRADED   SINCE   MESSAGE
baremetal   4.11.3   True        False         False      3d15h
----

. Delete the machine of the unhealthy member:
+
[source,terminal]
----

> $ oc delete machine -n openshift-machine-api clustername-8qw5l-master-2 <1>

Reviewer: naming

Author: Done, thank you!

----
+
If deletion of the machine is delayed for any reason or the command is obstructed and delayed, you can force deletion by removing the machine object finalizer field. 
+
[IMPORTANT]
====
Do not interrupt machine deletion with `control-c`. You must allow the command to proceed to completion. Open a new terminal window to edit and delete the finalizer fields.
====
+
[source,terminal]
----

> $ oc edit machine -n openshift-machine-api clustername-8qw5l-master-0

Reviewer: naming. And we are deleting control-plane-2, so finalizer should be deleted for him :)

----
+
.. Find and delete the fields:
+
[source,terminal]
----
finalizers:
- machine.machine.openshift.io
----
+
and save the change. 
+
[source,terminal]
----

> machine.machine.openshift.io/clustername8qw5l-master-0 edited

Reviewer: ditto

Author: Done, thank you!

examplecluster-control-plane-0         Running                          3h11m   openshift-control-plane-0   baremetalhost:///openshift-machine-api/openshift-control-plane-0/da1ebe11-3ff2-41c5-b099-0aa41222964e   externally provisioned <1>
examplecluster-control-plane-1         Running                          3h11m   openshift-control-plane-1   baremetalhost:///openshift-machine-api/openshift-control-plane-1/d9f9acbc-329c-475e-8d81-03b20280a3e1   externally provisioned
examplecluster-control-plane-2         Running                          3h11m   openshift-control-plane-2   baremetalhost:///openshift-machine-api/openshift-control-plane-2/3354bdac-61d8-410f-be5b-6a395b056135   externally provisioned
examplecluster-compute-0   Running                          165m    openshift-compute-0   baremetalhost:///openshift-machine-api/openshift-compute-0/3d685b81-7410-4bb3-80ec-13a31858241f   provisioned
examplecluster-compute-1   Running                          165m    openshift-compute-1   baremetalhost:///openshift-machine-api/openshift-compute-1/0fdae6eb-2066-4241-91dc-e7ea72ab13b9   provisioned 

> ----

Reviewer: tabbing for 2 last lines

Author: Done, thank you!

.Prerequisites

* You have identified the unhealthy bare metal etcd member.
* You have verified that either the machine is not running or the node is not ready.
* You have access to the cluster as a user with the `cluster-admin` role.
* You have taken an etcd backup.
+
[IMPORTANT]
====
It is important to take an etcd backup before performing this procedure so that your cluster can be restored if you encounter any issues.
====

.Procedure

. Verify and remove the unhealthy member.

.. Choose a pod that is _not_ on the affected node:
+
In a terminal that has access to the cluster as a `cluster-admin` user, run the following command:
+
[source,terminal]
----
$ oc get pods -n openshift-etcd -o wide | grep etcd | grep -v guard
----
+
.Example output
[source,terminal]
----
etcd-openshift-master-10                5/5     Running     11          3h56m   192.168.10.9   openshift-master-10   <none>           <none>
etcd-openshift-master-11                5/5     Running     0          3h54m   192.168.10.10   openshift-master-11   <none>           <none>

> etcd-openshift-master-12                5/5     Running     0          3h58m   192.168.10.11   openshift-master-12   <none>           <none>

Reviewer: why 10, 11, 12?

Author: The prior ones were in sequence, 0, 1, 2. I wanted to change the text somewhat, so to keep the sequence "correct", I chose simply to add 10 to each name.

+
[IMPORTANT]
====
Do not interrupt machine deletion with `control-c`. You must allow the command to proceed to completion. Open a new terminal window to edit and delete the finalizer fields.
====
+
[source,terminal]
----
$ oc edit machine -n openshift-machine-api clustername-8qw5l-master-0
----
+
.. Find and delete the fields:
+
[source,terminal]
----
finalizers:
- machine.machine.openshift.io
----
+
and save the change. 
+
[source,terminal]
----
machine.machine.openshift.io/clustername8qw5l-master-0 edited
----
+
.. Verify the machine was deleted:
+
[source,terminal]
----
$ oc get machines -n openshift-machine-api -o wide
----
+
.Example output
[source,terminal]
----
NAME                            PHASE     TYPE   REGION   ZONE   AGE     NODE                 PROVIDERID                                                                                       STATE
ocp-edge-4zg89-master-10         Running                          3h11m   openshift-master-0   baremetalhost:///openshift-machine-api/openshift-master-10/da1ebe11-3ff2-41c5-b099-0aa41222964e   externally provisioned

> ocp-edge-4zg89-master-11         Running                          3h11m   openshift-master-1   baremetalhost:///openshift-machine-api/openshift-master-1/d9f9acbc-329c-475e-8d81-03b20280a3e1   externally provisioned

Reviewer: please, change here as well to 0 and 1

Author: Thank you! Those have "tab" instead of "space" so my search missed those.

Author: Done! Also found another glitch and corrected it, so thank you, twice!

Reviewer: why I still see 10 and 11?

ocp-edge-4zg89-worker-0-8v2jj   Running                          165m    openshift-worker-0   baremetalhost:///openshift-machine-api/openshift-worker-0/3d685b81-7410-4bb3-80ec-13a31858241f   provisioned
ocp-edge-4zg89-worker-0-jb5z9   Running                          165m    openshift-worker-1   baremetalhost:///openshift-machine-api/openshift-worker-1/0fdae6eb-2066-4241-91dc-e7ea72ab13b9   provisioned
----

. Remove the old `BareMetalHost` object:
+
[source,terminal]
----
$ oc delete bmh openshift-master-2 -n openshift-machine-api

baremetalhost.metal3.io "openshift-master-12" deleted
----
+
After you remove the `BareMetalHost` and `Machine` objects, then the `Machine` controller automatically deletes the `Node` object.
+
If, after deletion of bmh, the machine node requires excessive time for deletion, the machine node can be deleted using: 
+
[source,terminal]
----
$ oc delete node openshift-master-12

> node "openshift-master-12" deleted

Author: Thank You! Done, please stand by.

secret "etcd-peer-openshift-master-2" deleted
----

... Delete the serving secret:
+
[source,terminal]
----
$ oc delete secret etcd-serving-metrics-openshift-master-2 -n openshift-etcd

secret "etcd-serving-metrics-openshift-master-2" deleted
----

... Delete the metrics secret:
+
[source,terminal]
----
$ oc delete secret etcd-serving-openshift-master-2 -n openshift-etcd

secret "etcd-serving-openshift-master-2" deleted
----

. Delete the control plane machine.
+
If you are running installer-provisioned infrastructure, or you used the Machine API to create your machines, follow these steps. Otherwise, you must create the new master using the same method that was used to originally create it.

.. Obtain the machine for the unhealthy member.
+
In a terminal that has access to the cluster as a `cluster-admin` user, run the following command:
+
[source,terminal]
----
$ oc get machines -n openshift-machine-api -o wide
----
+
.Example output
[source,terminal]
----
NAME                            PHASE     TYPE   REGION   ZONE   AGE     NODE                 PROVIDERID                                                                                       STATE

> ocp-edge-4zg89-master-0         Running                          3h11m   openshift-master-0   baremetalhost:///openshift-machine-api/openshift-master-10/da1ebe11-3ff2-41c5-b099-0aa41222964e   externally provisioned <1>

Reviewer: ```suggestion
examplecluster-control-plane-0         Running                          3h11m   openshift-master-0   baremetalhost:///openshift-machine-api/openshift-master-1/da1ebe11-3ff2-41c5-b099-0aa41222964e   externally provisioned <1>
```

Author: Done, thank you!

  - address: fe80::4adf:37ff:feb0:8aa1%ens1f1.373
    type: InternalDNS
  - address: fe80::4adf:37ff:feb0:8aa1%ens1f1.371
    type: Hostname
  lastUpdated: "2020-04-20T17:44:29Z"
  nodeRef:
    kind: Machine
    name: fe80::4adf:37ff:feb0:8aa1%ens1f1.372
    uid: acca4411-af0d-4387-b73e-52b2484295ad
  phase: Running
  providerStatus:
    apiVersion: machine.openshift.io/v1beta1
    conditions:
    - lastProbeTime: "2020-04-20T16:53:50Z"
      lastTransitionTime: "2020-04-20T16:53:50Z"
      message: machine successfully created
      reason: MachineCreationSucceeded
      status: "True"
      type: MachineCreation
    instanceId: i-0fdb85790d76d0c3f
    instanceState: stopped
    kind: Machine
----

. Change the `metadata.name` field to a new name.
+
It is recommended to keep the same base name as the old machine and change the ending number to the next available number. In this example, `clustername-8qw5l-master-0` is changed to `clustername-8qw5l-master-2`.
+
For example:
+
[source,yaml]
----
apiVersion: machine.openshift.io/v1beta1
kind: Machine
metadata:
  ...
  name: ocp-edge-4zg89-master-3
  ...
----

.. Remove the `spec.providerID` field:
+
[source,yaml]
----

>   providerID: baremetalhost:///openshift-machine-api/openshift-master-12/3354bdac-61d8-410f-be5b-6a395b056135

Reviewer: ```suggestion
  providerID: baremetalhost:///openshift-machine-api/openshift-control-plane-2/3354bdac-61d8-410f-be5b-6a395b056135
```

Author: Done, thank you!

. Delete the machine of the unhealthy member:
+
[source,terminal]
----
$ oc delete machine -n openshift-machine-api clustername-8qw5l-master-2 <1>
----
+
If deletion of the machine is delayed for any reason or the command is obstructed and delayed, you can force deletion by removing the machine object finalizer field. 
+
[IMPORTANT]
====
Do not interrupt machine deletion with `control-c`. You must allow the command to proceed to completion. Open a new terminal window to edit and delete the finalizer fields.
====
+
[source,terminal]
----
$ oc edit machine -n openshift-machine-api clustername-8qw5l-master-0
----
+
.. Find and delete the fields:
+
[source,terminal]
----
finalizers:
- machine.machine.openshift.io
----
+
and save the change. 
+
[source,terminal]
----
machine.machine.openshift.io/clustername8qw5l-master-0 edited
----
+
.. Verify the machine was deleted:
+
[source,terminal]
----
$ oc get machines -n openshift-machine-api -o wide
----
+
.Example output
[source,terminal]
----
NAME                            PHASE     TYPE   REGION   ZONE   AGE     NODE                 PROVIDERID                                                                                       STATE

> ocp-edge-4zg89-master-10         Running                          3h11m   openshift-master-0   baremetalhost:///openshift-machine-api/openshift-master-10/da1ebe11-3ff2-41c5-b099-0aa41222964e   externally provisioned

Reviewer: ```suggestion
clusterexaqmple-control-plane-0         Running                          3h11m   openshift-control-plane-0   baremetalhost:///openshift-machine-api/openshift-control-plane-0/da1ebe11-3ff2-41c5-b099-0aa41222964e   externally provisioned
```

Reviewer: the same for following rows

Author: Done, thank you!

ocp-edge-4zg89-master-11         Running                          3h11m   openshift-master-1   baremetalhost:///openshift-machine-api/openshift-master-1/d9f9acbc-329c-475e-8d81-03b20280a3e1   externally provisioned
ocp-edge-4zg89-worker-0-8v2jj   Running                          165m    openshift-worker-0   baremetalhost:///openshift-machine-api/openshift-worker-0/3d685b81-7410-4bb3-80ec-13a31858241f   provisioned
ocp-edge-4zg89-worker-0-jb5z9   Running                          165m    openshift-worker-1   baremetalhost:///openshift-machine-api/openshift-worker-1/0fdae6eb-2066-4241-91dc-e7ea72ab13b9   provisioned
----

. Remove the old `BareMetalHost` object:
+
[source,terminal]
----
$ oc delete bmh openshift-master-2 -n openshift-machine-api

> baremetalhost.metal3.io "openshift-master-12" deleted

Reviewer: 12 -> 2

Author: Done, thank you!

----
+
After you remove the `BareMetalHost` and `Machine` objects, then the `Machine` controller automatically deletes the `Node` object.
+
If, after deletion of bmh, the machine node requires excessive time for deletion, the machine node can be deleted using: 
+
[source,terminal]
----

> $ oc delete node openshift-master-12

Reviewer: ditto

Author: Done, thank you!

> node "openshift-master-12" deleted

Reviewer: ditto

Author: Done, thank you!

  bootMACAddress: 48:df:37:b0:8a:a0
  bootMode: UEFI
  externallyProvisioned: false
  online: true
  rootDeviceHints:
    deviceName: /dev/sda
  userData:
    name: master-user-data-managed
    namespace: openshift-machine-api
EOF
----
+
Note: username and password can be found from the other bmh's secrets. The protocol to use in bmc:address: can be taken from other bmh objects.
+
[IMPORTANT]
====
If you reuse the `BareMetalHost` object definition from an existing control plane host, do not leave the `externallyProvisioned` field set to `true`.

Existing control plane `BareMetalHost` objects may have the `externallyProvisioned` flag set to `true` if they were provisioned by the {product-title} installation program.
====
+
After the inspection is complete, the `BareMetalHost` object is created and available to be provisioned.

. Verify the creation process using available `BareMetalHost` objects:
+
[source,terminal]

----
$ oc get bmh -n openshift-machine-api

NAME STATE CONSUMER ONLINE ERROR AGE
openshift-master-0 externally provisioned ocp-edge-4zg89-master-0 true 4h48m
openshift-master-1 externally provisioned ocp-edge-4zg89-master-1 true 4h48m

> openshift-master-2 provisioned ocp-edge-4zg89-master-3 true 47m

Author: Done. Thank you!

openshift-worker-0 provisioned ocp-edge-4zg89-worker-0-8v2jj true 4h48m
openshift-worker-1 provisioned ocp-edge-4zg89-worker-0-jb5z9 true 4h48m

$ oc get machine -n openshift-machine-api

NAME PHASE TYPE REGION ZONE AGE
ocp-edge-4zg89-master-0 Running 4h49m
ocp-edge-4zg89-master-1 Running 4h48m
ocp-edge-4zg89-master-3 Running 25m
ocp-edge-4zg89-worker-0-8v2jj Running 4h23m
ocp-edge-4zg89-worker-0-jb5z9 Running 4h23m

$ oc get nodes

NAME STATUS ROLES AGE VERSION
openshift-master-0 Ready master 4h26m v1.24.0+9546431
openshift-master-1 Ready master 4h26m v1.24.0+9546431
openshift-master-2 Ready master 12m v1.24.0+9546431
openshift-worker-0 Ready worker 3h58m v1.24.0+9546431
openshift-worker-1 Ready worker 3h58m v1.24.0+9546431

> ----

Reviewer: Please, delete rows 458-476. On this stage no machine and no node are created yet, only bmh is added in the state available

Author: Done, thank you!

+
.. Create the new control plane machine using the `new-master-machine.yaml` file:
+
[source,terminal]
----
$ oc apply -f new-master-machine.yaml
----

.. Verify that the new machine has been created:
+
[source,terminal]
----
$ oc get machines -n openshift-machine-api -o wide
----
+
.Example output
[source,terminal]
----
NAME                            PHASE     TYPE   REGION   ZONE   AGE     NODE                 PROVIDERID                                                                                       STATE

> ocp-edge-4zg89-master-0         Running                          3h11m   openshift-master-0   baremetalhost:///openshift-machine-api/openshift-master-10/da1ebe11-3ff2-41c5-b099-0aa41222964e   externally provisioned <1>

Reviewer: as mentioned above

Author: Done, thank you!

+
[source,terminal]
----
$ oc get pods -n openshift-etcd -o wide | grep etcd | grep -v guard
----
+
.Example output
[source,terminal]
----
etcd-openshift-master-0                5/5     Running     0          105m
etcd-openshift-master-1                5/5     Running     0          107m
etcd-openshift-master-2                5/5     Running     0          103m 
----
+
If the output from the previous command only lists two pods, you can manually force an etcd redeployment. In a terminal that has access to the cluster as a `cluster-admin` user, run the following command:
+
[source,terminal]
----
$ oc patch etcd cluster -p='{"spec": {"forceRedeploymentReason": "recovery-'"$( date --rfc-3339=ns )"'"}}' --type=merge <1>
----
+
<1> The `forceRedeploymentReason` value must be unique, which is why a timestamp is appended.
+
.To verify there are exactly three etcd members, connect to the running etcd container, passing in the name of a pod that was not on the affected node:
+
.In a terminal that has access to the cluster as a `cluster-admin` user, run the following command:
+
[source,terminal]
----

> $ oc rsh -n openshift-etcd etcd-openshift-master-10

Reviewer: 10 -> 0

Author: Done. Thank you!

// Module included in the following assemblies:
//
// * /backup_and_restore/control_plane_backup_and_restore/replacing-unhealthy-etcd-member.adoc

:_content-type: PROCEDURE
[id="restore-replace-stopped-baremetal-etcd-member_{context}"]
= Replacing an unhealthy bare metal etcd member whose machine is not running or whose node is not ready

This procedure details the steps to replace a bare metal etcd member that is unhealthy either because the machine is not running or because the node is not ready.

> If you are running installer-provisioned infrastructure or you used the Machine API to create your machines, follow these steps. Otherwise you must create the new master using the same method that was used to originally create it. 

Author: Excellent catch, thank you, yes!

.Prerequisites

* You have identified the unhealthy bare metal etcd member.
* You have verified that either the machine is not running or the node is not ready.
* You have access to the cluster as a user with the `cluster-admin` role.
* You have taken an etcd backup.
+
[IMPORTANT]
====

> It is important to take an etcd backup before performing this procedure so that your cluster can be restored if you encounter any issues.

Author: VERY glad to make that change!

+
[source,terminal]
----
$ oc rsh -n openshift-etcd etcd-openshift-control-plane-0
----

.. View the member list:
+
[source,terminal]
----
sh-4.2# etcdctl member list -w table
----
+
.Example output
[source,terminal]
----
+------------------+---------+--------------------+---------------------------+---------------------------+---------------------+
| ID               | STATUS  | NAME                      | PEER ADDRS                  | CLIENT ADDRS                | IS LEARNER |
+------------------+---------+--------------------+---------------------------+---------------------------+---------------------+
| 7a8197040a5126c8 | started | openshift-control-plane-2 | https://192.168.10.11:2380/ | https://192.168.10.11:2379/ | false |
| 8d5abe9669a39192 | started | openshift-control-plane-1 | https://192.168.10.10:2380/ | https://192.168.10.10:2379/ | false |
| cc3830a72fc357f9 | started | openshift-control-plane-0 | https://192.168.10.9:2380/ | https://192.168.10.9:2379/   | false |
+------------------+---------+--------------------+---------------------------+---------------------------+---------------------+
----
+

> Take note of the ID and the name of the unhealthy etcd member, because these values are required later in the procedure. The `$ etcdctl endpoint health` command will list the removed member until the replacement procedure is completed and the new member is added.

Author: Very good catch - that was a cut/paste error. Thank you!

----

.. View the member list again and verify that the member was removed:
+
[source,terminal]
----
sh-4.2# etcdctl member list -w table
----
+
.Example output
[source,terminal]
----
+------------------+---------+--------------------+---------------------------+---------------------------+-------------------------+
| ID               | STATUS  | NAME                      | PEER ADDRS                  | CLIENT ADDRS                | IS LEARNER |
+------------------+---------+--------------------+---------------------------+---------------------------+-------------------------+
| 7a8197040a5126c8 | started | openshift-control-plane-2 | https://192.168.10.11:2380/ | https://192.168.10.11:2379/ | false |
| 8d5abe9669a39192 | started | openshift-control-plane-1 | https://192.168.10.10:2380/ | https://192.168.10.10:2379/ | false |
+------------------+---------+--------------------+---------------------------+---------------------------+-------------------------+
----
+
You can now exit the node shell.
+
[IMPORTANT]
====
After you remove the member, the cluster might be unreachable for a short time while the remaining etcd instances reboot.
====

> . Remove the old secrets for the unhealthy etcd member that was removed.

Reviewer 2: The substeps in this part of the procedure need the "by running the following command" or [similar explicit instructions](https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#code-blocks-command-syntax-and-example-output) appended to it

Author: Done - good point!

----
Pass in the name of the unhealthy etcd member that you took note of earlier in this procedure.
+
There is a peer, serving, and metrics secret as shown in the following output:
+

[source,terminal]
----
etcd-peer-openshift-control-plane-2             kubernetes.io/tls   2   134m
etcd-serving-metrics-openshift-control-plane-2  kubernetes.io/tls   2   134m
etcd-serving-openshift-control-plane-2          kubernetes.io/tls   2   134m
----

.. Delete the secrets for the unhealthy etcd member that was removed.

... Delete the peer secret:
+
[source,terminal]
----
$ oc delete secret etcd-peer-openshift-control-plane-2 -n openshift-etcd

secret "etcd-peer-openshift-control-plane-2" deleted
----

... Delete the serving secret:
+
[source,terminal]
----
$ oc delete secret etcd-serving-metrics-openshift-control-plane-2 -n openshift-etcd

secret "etcd-serving-metrics-openshift-control-plane-2" deleted
----

... Delete the metrics secret:
+
[source,terminal]
----
$ oc delete secret etcd-serving-openshift-control-plane-2 -n openshift-etcd

secret "etcd-serving-openshift-control-plane-2" deleted
----

. Delete the control plane machine.
+

> If you are running installer-provisioned infrastructure, or you used the Machine API to create your machines, follow these steps. Otherwise, you must create the new master using the same method that was used to originally create it.

Reviewer 2: Similar to an earlier comment: is there a way we can avoid using the term "master"?

examplecluster-control-plane-2    Running                          3h11m   openshift-control-plane-2   baremetalhost:///openshift-machine-api/openshift-control-plane-2/3354bdac-61d8-410f-be5b-6a395b056135   externally provisioned
examplecluster-compute-0          Running                          165m    openshift-compute-0         baremetalhost:///openshift-machine-api/openshift-compute-0/3d685b81-7410-4bb3-80ec-13a31858241f         provisioned
examplecluster-compute-1          Running                          165m    openshift-compute-1         baremetalhost:///openshift-machine-api/openshift-compute-1/0fdae6eb-2066-4241-91dc-e7ea72ab13b9         provisioned
----
<1> This is the control plane machine for the unhealthy node, `examplecluster-control-plane-2`.

.. Save the machine configuration to a file on your file system:
+
[source,terminal]
----
$ oc get machine examplecluster-control-plane-2 \ <1>
    -n openshift-machine-api \
    -o yaml \
    > new-master-machine.yaml
----

> Specify the name of the control plane machine for the unhealthy node.

Reviewer 2: Did you mean for this to be the callout? E.g.
```suggestion
<1> Specify the name of the control plane machine for the unhealthy node.
```

status:
  addresses:
  - address: ""
    type: InternalIP
  - address: fe80::4adf:37ff:feb0:8aa1%ens1f1.373
    type: InternalDNS
  - address: fe80::4adf:37ff:feb0:8aa1%ens1f1.371
    type: Hostname
  lastUpdated: "2020-04-20T17:44:29Z"
  nodeRef:
    kind: Machine
    name: fe80::4adf:37ff:feb0:8aa1%ens1f1.372
    uid: acca4411-af0d-4387-b73e-52b2484295ad
  phase: Running
  providerStatus:
    apiVersion: machine.openshift.io/v1beta1
    conditions:
    - lastProbeTime: "2020-04-20T16:53:50Z"
      lastTransitionTime: "2020-04-20T16:53:50Z"
      message: machine successfully created
      reason: MachineCreationSucceeded
      status: "True"
      type: MachineCreation
    instanceId: i-0fdb85790d76d0c3f
    instanceState: stopped
    kind: Machine
----

. Change the `metadata.name` field to a new name.
+
It is recommended to keep the same base name as the old machine and change the ending number to the next available number. In this example, `examplecluster-control-plane-2` is changed to `examplecluster-control-plane-3`.
+
For example:
+
[source,yaml]
----
apiVersion: machine.openshift.io/v1beta1
kind: Machine
metadata:
  ...
  name: examplecluster-control-plane-3
  ...
----

> .. Remove the `spec.providerID` field:

Reviewer 2: Are these substeps? They seem disconnected from the originating step. Maybe a way to approach this whole step is to have the parent step be "Change or remove the following fields", and have the substep start with what is currently step 4. What do you think?

Reviewer 2: @GroceryBoyJr A note sounds like a good idea! 👍

+
[source,yaml]
----
  providerID: baremetalhost:///openshift-machine-api/openshift-control-plane-2/3354bdac-61d8-410f-be5b-6a395b056135
----

.. Remove the `metadata.annotations` and `metadata.generation` fields:
+
[source,yaml]
----
  annotations:
    machine.openshift.io/instance-state: externally provisioned
  ...
  generation: 2
----

> .. Remove the `spec.conditions`, `spec.lastUpdated`, `spec.nodeRef` and `spec.phase' fields:

Reviewer 2: ```suggestion
.. Remove the `spec.conditions`, `spec.lastUpdated`, `spec.nodeRef` and `spec.phase` fields:
```

Author: This is where I found a "hidden backtick". I THINK it's a vscode/codium thing (trying to "close" the backticks) but I'm not sure at this point. Will take a look at the file with vim to double-check.

+
[source,yaml]
----
  lastTransitionTime: "2022-08-03T08:40:36Z"
message: 'Drain operation currently blocked by: [{Name:EtcdQuorumOperator Owner:clusteroperator/etcd}]'
reason: HookPresent
severity: Warning
status: "False"

type: Drainable
lastTransitionTime: "2022-08-03T08:39:55Z"
status: "True"
type: InstanceExists

lastTransitionTime: "2022-08-03T08:36:37Z"
status: "True"
type: Terminable
lastUpdated: "2022-08-03T08:40:36Z"
nodeRef:
kind: Node
name: openshift-control-plane-2
uid: 788df282-6507-4ea2-9a43-24f237ccbc3c
phase: Running
----

> . Ensure that the Bare Metal Operator is available:

Reviewer 2: ```suggestion
. Ensure that the Bare Metal Operator is available by running the following command:
```

Author: Agreed, thank you.

+
[source,terminal]
----
$ oc get clusteroperator baremetal
----
+
.Example output
[source,terminal]
----
NAME        VERSION   AVAILABLE   PROGRESSING   DEGRADED   SINCE   MESSAGE
baremetal   4.11.3    True        False         False      3d15h
----

> . Delete the machine of the unhealthy member:

Reviewer 2: ```suggestion
. Delete the machine of the unhealthy member by running the following command:
```

Author: I went with "using this command" to try to be a little less repetitive, but completely agree with your suggestion!

+
[source,terminal]
----
$ oc delete machine -n openshift-machine-api examplecluster-control-plane-2 <1>
----
+

> If deletion of the machine is delayed for any reason or the command is obstructed and delayed, you can force deletion by removing the machine object finalizer field. 

Reviewer 2: Did you mean for this to be a callout? E.g.
```suggestion
<1> If deletion of the machine is delayed for any reason or the command is obstructed and delayed, you can force deletion by removing the machine object finalizer field. 
```

+
[IMPORTANT]
====

> Do not interrupt machine deletion with `control-c`. You must allow the command to proceed to completion. Open a new terminal window to edit and delete the finalizer fields.

Reviewer 2: ISG nit:
```suggestion
Do not interrupt machine deletion by pressing `Ctrl+c`. You must allow the command to proceed to completion. Open a new terminal window to edit and delete the finalizer fields.
```

Author: Oh dang! Thank you!

====
+

> [source,terminal]

Reviewer 2: Is this code block a duplicate?

Author: No, the first one is the primary command, the second one is a follow-on command to delete the finalizer fields.

----
$ oc edit machine -n openshift-machine-api examplecluster-control-plane-2
----
+
.. Find and delete the fields:
+
[source,terminal]
----
finalizers:
- machine.machine.openshift.io
----
+

> and save the change. 

Reviewer 2: ```suggestion
Save your changes: 
```

Author: Brilliant, thank you.

+
[source,terminal]
----
machine.machine.openshift.io/examplecluster-control-plane-2 edited
----
+

> .. Verify the machine was deleted:

Reviewer 2: ```suggestion
.. Verify the machine was deleted by running the following command:
```

Author: Yes! Done.

----
+
.Example output
[source,terminal]
----
NAME                              PHASE     TYPE   REGION   ZONE   AGE     NODE                                 PROVIDERID                                                                                       STATE
examplecluster-control-plane-0    Running                          3h11m   openshift-control-plane-0   baremetalhost:///openshift-machine-api/openshift-control-plane-0/da1ebe11-3ff2-41c5-b099-0aa41222964e   externally provisioned
examplecluster-control-plane-1    Running                          3h11m   openshift-control-plane-1   baremetalhost:///openshift-machine-api/openshift-control-plane-1/d9f9acbc-329c-475e-8d81-03b20280a3e1   externally provisioned
examplecluster-compute-0          Running                          165m    openshift-compute-0         baremetalhost:///openshift-machine-api/openshift-compute-0/3d685b81-7410-4bb3-80ec-13a31858241f         provisioned
examplecluster-compute-1          Running                          165m    openshift-compute-1         baremetalhost:///openshift-machine-api/openshift-compute-1/0fdae6eb-2066-4241-91dc-e7ea72ab13b9         provisioned
----

> . Remove the old `BareMetalHost` object:

Reviewer 2: ```suggestion
. Remove the old `BareMetalHost` object by running the following command:
```

Author: Yes! Varied slightly to avoid repetition.

+
[source,terminal]
----
$ oc delete bmh openshift-control-plane-2 -n openshift-machine-api

> baremetalhost.metal3.io "openshift-control-plane-2" deleted

Reviewer 2: I think this should be separated into an example output

Author: Agreed. Done!

----
+
After you remove the `BareMetalHost` and `Machine` objects, then the `Machine` controller automatically deletes the `Node` object.
+

> If, after deletion of bmh, the machine node requires excessive time for deletion, the machine node can be deleted using: 

Reviewer 2: ```suggestion
If, after deletion of the `BareMetalHost` object, the machine node requires excessive time for deletion, the machine node can be deleted using: 
```

Author: Thank you, yes.

openshift-control-plane-0 Ready master 3h24m v1.24.0+9546431
openshift-control-plane-1 Ready master 3h24m v1.24.0+9546431
openshift-compute-0       Ready worker 176m v1.24.0+9546431
openshift-compute-1       Ready worker 176m v1.24.0+9546431
----

. Create the new `BareMetalHost` object and the secret to store the BMC credentials:

+
[source,terminal]
----
$ cat <<EOF | oc apply -f -
apiVersion: v1
kind: Secret
metadata:
  name: openshift-control-plane-2-bmc-secret
  namespace: openshift-machine-api
data:
  password: <password>
  username: <username>
type: Opaque
---
apiVersion: metal3.io/v1alpha1
kind: BareMetalHost
metadata:
  name: openshift-control-plane-2
  namespace: openshift-machine-api
spec:
  automatedCleaningMode: disabled
  bmc:
    address: redfish://10.46.61.18:443/redfish/v1/Systems/1
    credentialsName: openshift-control-plane-2-bmc-secret
    disableCertificateVerification: true
  bootMACAddress: 48:df:37:b0:8a:a0
  bootMode: UEFI
  externallyProvisioned: false
  online: true
  rootDeviceHints:
    deviceName: /dev/sda
  userData:
    name: master-user-data-managed
    namespace: openshift-machine-api
EOF
----
+

> Note: username and password can be found from the other bmh's secrets. The protocol to use in bmc:address: can be taken from other bmh objects.

Reviewer 2: Is "bmh" a commonly understood acronym? If not, then suggestion to expand it out to bare metal host:
```suggestion
[NOTE]
====
The username and password can be found from the other bare metal host's secrets. The protocol to use in the `bmc.address` field can be taken from other Bare Metal Host objects.
====
```

Author: It is, but it's a specific modifier used in the command itself. Pulling that term out into the descriptions may affect readability to the point that I think I should go with your changes.  Also, SERIOUSLY agreed with you about changing the Note: to [NOTE]. I was afraid of repetition here but I think the change is a big improvement in consistency.

+
[IMPORTANT]
====
If you reuse the `BareMetalHost` object definition from an existing control plane host, do not leave the `externallyProvisioned` field set to `true`.

Existing control plane `BareMetalHost` objects may have the `externallyProvisioned` flag set to `true` if they were provisioned by the {product-title} installation program.
====
+
After the inspection is complete, the `BareMetalHost` object is created and available to be provisioned.

. Verify the creation process using available `BareMetalHost` objects:
+
[source,terminal]

----
$ oc get bmh -n openshift-machine-api

NAME                      STATE                  CONSUMER                      ONLINE ERROR   AGE
openshift-control-plane-0 externally provisioned examplecluster-control-plane-0 true         4h48m
openshift-control-plane-1 externally provisioned examplecluster-control-plane-1 true         4h48m
openshift-control-plane-2 available              examplecluster-control-plane-3 true         47m
openshift-compute-0       provisioned            examplecluster-compute-0       true         4h48m
openshift-compute-1       provisioned            examplecluster-compute-1       true         4h48m
----
+

> .. Create the new control plane machine using the `new-master-machine.yaml` file:

Reviewer 2: These substeps seem disconnected from the parent step. Should they be their own steps?

Author: Hmmm. i'm pondering that. I wanted the a,b,c,d to be "components of the 9th step". But that might be a bit complex? I'm thinking on it.....

examplecluster-control-plane-2         Running                          3h11m   openshift-control-plane-2   baremetalhost:///openshift-machine-api/openshift-control-plane-2/3354bdac-61d8-410f-be5b-6a395b056135   externally provisioned
examplecluster-compute-0               Running                          165m    openshift-compute-0         baremetalhost:///openshift-machine-api/openshift-compute-0/3d685b81-7410-4bb3-80ec-13a31858241f         provisioned
examplecluster-compute-1               Running                          165m    openshift-compute-1         baremetalhost:///openshift-machine-api/openshift-compute-1/0fdae6eb-2066-4241-91dc-e7ea72ab13b9         provisioned 
----
<1> The new machine, `clustername-8qw5l-master-3` is being created and is ready after the phase changes from `Provisioning` to `Running`.
+
It should take a few minutes for the new machine to be created. The etcd cluster Operator will automatically sync when the machine or node returns to a healthy state.

> .. Verify that bmh becomes provisioned and no error reported:

Reviewer 2: ```suggestion
.. Verify that the bare metal host becomes provisioned and no error reported by running the following command:
```

Author: Done, thank you!

+
[source,terminal]
----
$ oc get bmh -n openshift-machine-api
----
+
.Example output
[source,terminal]
----
$ oc get bmh -n openshift-machine-api
NAME                      STATE                  CONSUMER                       ONLINE ERROR AGE
openshift-control-plane-0 externally provisioned examplecluster-control-plane-0 true         4h48m
openshift-control-plane-1 externally provisioned examplecluster-control-plane-1 true         4h48m
openshift-control-plane-2 provisioned            examplecluster-control-plane-3 true          47m
openshift-compute-0       provisioned            examplecluster-compute-0       true         4h48m
openshift-compute-1       provisioned            examplecluster-compute-1       true         4h48m
----

> .. The new node is added and in ready state:

Reviewer 2: Is this a verification step?
```suggestion
.. Verify that the new node is added and in a ready state by running the following command:
```

Author: Done, YES! Thank You!

openshift-compute-1       Ready worker 3h58m v1.24.0+9546431
----

.Verification

. Verify that all etcd pods are running properly.
+
In a terminal that has access to the cluster as a `cluster-admin` user, run the following command:
+
[source,terminal]
----
$ oc get pods -n openshift-etcd -o wide | grep etcd | grep -v guard
----
+
.Example output
[source,terminal]
----
etcd-openshift-control-plane-0      5/5     Running     0     105m
etcd-openshift-control-plane-1      5/5     Running     0     107m
etcd-openshift-control-plane-2      5/5     Running     0     103m 
----
+
If the output from the previous command only lists two pods, you can manually force an etcd redeployment. In a terminal that has access to the cluster as a `cluster-admin` user, run the following command:
+
[source,terminal]
----
$ oc patch etcd cluster -p='{"spec": {"forceRedeploymentReason": "recovery-'"$( date --rfc-3339=ns )"'"}}' --type=merge <1>
----
+
<1> The `forceRedeploymentReason` value must be unique, which is why a timestamp is appended.
+
.To verify there are exactly three etcd members, connect to the running etcd container, passing in the name of a pod that was not on the affected node:
+

> .In a terminal that has access to the cluster as a `cluster-admin` user, run the following command:

Reviewer 2: ```suggestion
. In a terminal that has access to the cluster as a `cluster-admin` user, run the following command:
```

Author: What...the...hale. I have corrected this. It was a small paragraph of direction that had gotten munged up with too many "+" signs....THANK YOU SO MUCH.

+
[source,terminal]
----
$ oc rsh -n openshift-etcd etcd-openshift-control-plane-0
----
+
. View the member list:
+
[source,terminal]
----
sh-4.2# etcdctl member list -w table
----
+
.Example output
[source,terminal]
----
+------------------+---------+--------------------+---------------------------+---------------------------+-----------------+
|        ID        | STATUS  |        NAME        |        PEER ADDRS         |       CLIENT ADDRS        |    IS LEARNER    |
+------------------+---------+--------------------+---------------------------+---------------------------+-----------------+
| 7a8197040a5126c8 | started | openshift-control-plane-2 | https://192.168.10.11:2380 | https://192.168.10.11:2379 |   false |
| 8d5abe9669a39192 | started | openshift-control-plane-1 | https://192.168.10.10:2380 | https://192.168.10.10:2379 |   false |
| cc3830a72fc357f9 | started | openshift-control-plane-0 | https://192.168.10.9:2380 | https://192.168.10.9:2379 |     false |
+------------------+---------+--------------------+---------------------------+---------------------------+-----------------+
----
+

> If the output from the previous command lists more than three etcd members, you must carefully remove the unwanted member.

Reviewer 2: Suggestion to turn this into a note:
```suggestion
[NOTE]
====
If the output from the previous command lists more than three etcd members, you must carefully remove the unwanted member.
====
```

Author: agreed! Thank you!

+

> . Verify that all etcd members are healthy:

Reviewer 2: ```suggestion
. Verify that all etcd members are healthy by running the following command:
```

Author: WHO DOES THAT?!?! nOt mE of cOuRSe!

+
[source,terminal]
----

> # edctl endpoint health --cluster

Reviewer 2: Did you mean `etcdctl` here?

Author: Fatigue. That's my only reasonable answer here. Just fatigue. DOH!

----
+
.Example output
[source,terminal]
----
https://192.168.10.10:2379 is healthy: successfully committed proposal: took = 8.973065ms
https://192.168.10.9:2379 is healthy: successfully committed proposal: took = 11.559829ms
https://192.168.10.11:2379 is healthy: successfully committed proposal: took = 11.665203ms
----
+

> . Validate that all nodes are at the latest revision

Reviewer 2: ```suggestion
. Validate that all nodes are at the latest revision by running the following command:
```

Author: Butofcourse! Thank You!
